commit e99534d263435e270018451b0def2153e17d7caf (refs/changes/76/3576/3)
Author: Marsell Kukuljevic <marsell@joyent.com>
Date:   2018-03-13T12:20:54+00:00 (1 year, 7 months ago)
    
    TRITON-168 Regex anti-affinity rules fail unexpectedly
    Reviewed by: Pedro P. Candel <pedro@joyent.com>
    Approved by: Pedro P. Candel <pedro@joyent.com>

diff --git a/lib/machines.js b/lib/machines.js
index f1c548a..a97b05c 100644
--- a/lib/machines.js
+++ b/lib/machines.js
@@ -735,7 +735,12 @@ function getCreateOptions(req) {
             throw new InvalidArgumentError(
                 'cannot specify both "locality" (deprecated) and "affinity"');
         }
-        opts.affinity = params.affinity;
+
+        if (Array.isArray(params.affinity)) {
+            opts.affinity = params.affinity;
+        } else {
+            opts.affinity = [params.affinity];
+        }
     }
 
     Object.keys(pkg).forEach(function (p) {
diff --git a/lib/triton-affinity.js b/lib/triton-affinity.js
index 44c5aa3..7ee9036 100644
--- a/lib/triton-affinity.js
+++ b/lib/triton-affinity.js
@@ -428,7 +428,7 @@ function _vmsFromRule(opts, cb) {
             return;
         }
         opts.vmapi.listVms({
-            fields: 'uuid,alias,internal_metadata,docker',
+            fields: 'uuid,alias,server_uuid,internal_metadata,docker',
             owner_uuid: opts.ownerUuid,
             state: 'active'
         }, {
diff --git a/test/machines/affinity.js b/test/machines/affinity.js
index 5aa9d6b..3fc934b 100644
--- a/test/machines/affinity.js
+++ b/test/machines/affinity.js
@@ -52,6 +52,19 @@ function (suite, client, other, imgUuid, pkgUuid, headnodeUuid, cb) {
     });
 
 
+    // Same as above, but using a regular expression
+    suite.test('CreateMachine with affinity "container==/^' + CONTAINER_PREFIX +
+        '/"', function (t) {
+
+        var args = createArgs('container==/^' + CONTAINER_PREFIX + '/');
+
+        client.post('/my/machines', args, function (err, req, res, body) {
+            t.ok(err, 'VM with false affinity should fail');
+            t.end();
+        });
+    });
+
+
     // This should work: no container with name 'sdccloudapitest_affinity_*'.
     // This behaviour was changed in DAPI-306.
     suite.test('CreateMachine with affinity "container!=' + CONTAINER_PREFIX +
